### PR TITLE
[PLAT-349] Timeout of a Game

### PR DESCRIPTION
**Pull request by DidacSF**
## Description

This PR aims to tackle the problem described in https://ajunanetwork.atlassian.net/browse/PLAT-349 in order to do the following changes have been implemented:

* Added `on_idle` hook to the `pallet_ajuna_board` Pallet which checks periodically for stale Games to clean
* Added the `get_next_player` method to the `TurnBasedGame` trait
* Added the `abort` method to the `TurnBasedGame` trait

## Type of changes

- [ ] `build`: Changes that affect the build system or external dependencies (eg, Cargo, Docker)
- [ ] `ci`: Changes to CI configuration
- [ ] `docs`: Changes to documentation only
- [x] `feat`: Changes to add a new feature
- [ ] `fix`: Changes to fix a bug
- [x] `refactor`: Changes that do not alter functionality
- [ ] `style`: Changes to format the code
- [ ] `test`: Changes to add missing tests or correct existing tests

## Checklist

- [x] Tests for the changes have been added
- [x] Necessary documentation is added (if appropriate)
- [x] Formatted with `cargo fmt --all`
- [x] Linted with `cargo clippy --all-features -- -D warnings`
- [ ] Tested with `cargo test --workspace --all-features`

### CODE REVIEW

File: pallets/ajuna-board/src/lib.rs [hunk 1]
```diff
@@ -74,6 +77,14 @@ pub mod pallet {
 		/// Maximum number of players
 		#[pallet::constant]
 		type MaxNumberOfPlayers: Get<u32>;
+
+		#[pallet::constant]
+		/// Maximum number of successive blocks with no player activity
+		type MaxNumberOfIdleBlocks: Get<u32>;
```
Comment (cowboy-bebug) [inline pallets/ajuna-board/src/lib.rs:83]: Just a tiny nitpick :smile: 
```suggestion
		/// Maximum number of successive blocks with no player activity
		#[pallet::constant]
		type MaxNumberOfIdleBlocks: Get<u32>;
```

File: pallets/ajuna-board/src/lib.rs [hunk 2]
```diff
@@ -243,4 +287,31 @@ impl<T: Config> Pallet<T> {
...
+		board_state: &<<T as Config>::Game as TurnBasedGame>::State,
+	) {
+		// Cache result in storage, this would be cleared on `flush_winner`
+		BoardWinners::<T>::insert(board_id, winner.clone());
+		BoardExpiries::<T>::remove(board_id);
+		Self::seed_for_next(board_state);
+		Self::deposit_event(Event::GameFinished { board_id: *board_id, winner: winner.clone() });
+	}
+
+	fn set_or_update_expiry(board_id: T::BoardId) {
```
Comment (cowboy-bebug) [inline pallets/ajuna-board/src/lib.rs:303]: Could we rename this to:
```suggestion
	fn upsert_expiry(board_id: T::BoardId) {
```

File: pallets/ajuna-board/src/lib.rs [hunk 3]
```diff
@@ -243,4 +287,31 @@ impl<T: Config> Pallet<T> {
...
+
+		let block_of_expiry = current_block_number + expiry_block_count.into();
+
+		if BoardExpiries::<T>::contains_key(board_id) {
+			BoardExpiries::<T>::mutate(board_id, |entry| {
+				*entry = block_of_expiry;
+			});
+		} else {
+			BoardExpiries::<T>::insert(board_id, block_of_expiry);
+		}
```
Comment (cowboy-bebug) [inline pallets/ajuna-board/src/lib.rs:315]: Nice checks! But these lines can simply be:
```suggestion
		BoardExpiries::<T>::insert(board_id, block_of_expiry);
```
(`insert()` will mutate / replace the key with a new value)

File: pallets/ajuna-board/src/tests.rs [hunk 4]
```diff
@@ -176,3 +179,198 @@ fn should_finish_game_and_not_allow_new_game() {
...
+		);
+
+		// Game should expire on this block during on_idle
+		run_to_block(game_expiry);
+
+		// We play a turn extending it's expiry
+		assert_ok!(AjunaBoard::play_turn(Origin::signed(BOB), 1_u32));
+
+		// We run on_idle with the remaining weight
+
```
Comment (cowboy-bebug) [inline pallets/ajuna-board/src/tests.rs:254]: Nitpick:
```suggestion
```

**Review by cowboy-bebug — APPROVED** [review]
Looks great apart from minor nitpicks!
Nice work :shipit:

File: pallets/ajuna-board/src/lib.rs (hunk 3 above)
Comment (DidacSF) [reply pallets/ajuna-board/src/lib.rs:315]: Oh! I didn't know that, the docs unfortunately didn't mention such behavior, good to know!

**Pull request merged**